vendors/ti/boards/cc3220_launchpad/ports/common_io/iot_uart.c

```diff
@@ -0,0 +1,478 @@
+    {
+        return IOT_UART_INVALID_VALUE;
+    }
+
+    int32_t ret = IOT_UART_SUCCESS;
+    const UARTCC32XX_HWAttrsV1 * hwAttrs = _getHWAttrs( uartHandle->tiUartHandle );
+
+    ClockP_FreqHz freq;
+    ClockP_getCpuFreq( &freq );
+
+    switch( uartRequest )
+    {
+        case eUartSetConfig:
+           {
+               if( _isBusy( uartHandle->tiUartHandle ) )
+               {
+                   ret = IOT_UART_BUSY;
+                   break;
+               }
+
+               IotUARTConfig_t * pConfig = ( IotUARTConfig_t * ) pBuffer;
+               unsigned long tiFlowMode = pConfig->ucFlowControl == 0 ? UART_FLOWCONTROL_NONE : UART_FLOWCONTROL_TX | UART_FLOWCONTROL_RX;
+               unsigned long tiConfig = dataLength[ pConfig->ucWordlength - 5 ] | stopBits[ pConfig->xStopbits ] | parityType[ pConfig->xParity ];
+
+               MAP_UARTFlowControlSet( hwAttrs->baseAddr, tiFlowMode );
+               MAP_UARTConfigSetExpClk( hwAttrs->baseAddr, freq.lo, pConfig->ulBaudrate, tiConfig );
+
+               break;
+           }
+
+        case eUartGetConfig:
+           {
+               IotUARTConfig_t * pConfig = ( IotUARTConfig_t * ) pBuffer;
+               unsigned long baud = 0;
+               unsigned long tiConfig = 0;
+               unsigned long tiFlowMode = 0;
+
+               tiFlowMode = MAP_UARTFlowControlGet( hwAttrs->baseAddr );
+               MAP_UARTConfigGetExpClk( hwAttrs->baseAddr, freq.lo, &baud, &tiConfig );
+
+               pConfig->ucFlowControl = tiFlowMode == UART_FLOWCONTROL_NONE ? 0 : 1;
+               pConfig->ulBaudrate = baud;
+               pConfig->ucWordlength = ( uint8_t ) _indexOf( dataLength, sizeof( dataLength ), ( tiConfig & UART_CONFIG_WLEN_MASK ) ) + 5;
+               pConfig->xStopbits = ( IotUARTStopBits_t ) _indexOf( stopBits, sizeof( stopBits ), ( tiConfig & UART_CONFIG_STOP_MASK ) );
+               pConfig->xParity = ( IotUARTParity_t ) _indexOf( parityType, sizeof( parityType ), ( tiConfig & UART_CONFIG_PAR_MASK ) );
+
+               break;
+           }
+
+        case eGetTxNoOfbytes:
+            *( uint16_t * ) pBuffer = uartHandle->txNoOfbytes;
+            break;
+
+        case eGetRxNoOfbytes:
+            *( uint16_t * ) pBuffer = uartHandle->rxNoOfbytes;
+            break;
+
+        default:
+            ret = IOT_UART_INVALID_VALUE;
+            break;
+    }
+
+    return ret;
+}
+
+/*-----------------------------------------------------------*/
+
+int32_t iot_uart_cancel( IotUARTHandle_t const uartHandle )
+{
+    if( ( uartHandle == NULL ) || ( uartHandle->tiUartHandle == NULL ) )
+    {
+        return IOT_UART_INVALID_VALUE;
+    }
+
+    if( _isIdle( uartHandle->tiUartHandle ) )
+    {
+        return IOT_UART_NOTHING_TO_CANCEL;
+    }
+
+    UART_writeCancel( uartHandle->tiUartHandle );
+    UART_readCancel( uartHandle->tiUartHandle );
+
+    /* Have to sleep for a short period to let cancel finish. */
+    vTaskDelay( pdMS_TO_TICKS( 1000 ) );
```

Comment:
I think it would be better to use a semaphore here. Waiting 1 second is pretty long. Reading the TI API, when you call cancel it invokes the registered callback (whether it be read or write). You could probably add some magic number or something to the callback, signaling you to release the semaphore in the callback.

Reply:
Maybe setting some variable in the uartHandle

Reply:
+1, also hard-coded delays in common code like this are usually a mistake.  If you can't have a structure member as @lundinc2 suggests, then at least use a `#define`.